### PR TITLE
ZK-4914: Add toggleable debug / error trace for Clients EvalJavascript

### PR DESCRIPTION
**Pull request by DevChu**

### CODE REVIEW

File: zk/src/archive/web/js/zk/mount.ts [hunk 1]
```diff
@@ -837,6 +837,12 @@ jq(function () {
 	.on((document.hidden !== undefined ? '' : zk.vendor_) + 'visibilitychange', function (evt) {
 		zAu._onVisibilityChange();
 	});
+
+	if (zk['scriptErrorHandlerEnabled']) {
```
Comment (DevChu) [inline zk/src/archive/web/js/zk/mount.ts:841]: d.ts

File: zk/src/archive/web/js/zk/au.ts [hunk 2]
```diff
@@ -1312,7 +1312,16 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
 	 */
 	script: function (script) {
-		jq.globalEval(script);
+		// eslint-disable-next-line no-useless-catch
+		let scriptErrorHandler = zk['scriptErrorHandler'];
+		if (scriptErrorHandler)
+			jq(window).on('error', {script: script}, scriptErrorHandler);
+		try {
+			jq.globalEval(script);
+		} finally {
```
Comment (DevChu) [inline zk/src/archive/web/js/zk/au.ts:1321]: why no catch 
https://bugs.jquery.com/ticket/8784

File: zk/src/org/zkoss/zk/ui/script/DefaultScriptErrorListener.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * A Default Event Listener for handling client script error
+ * @author jameschu
+ */
+public class DefaultScriptErrorListener implements EventListener {
+	private static final Logger log = LoggerFactory.getLogger(DefaultScriptErrorListener.class);
+
+	@Override
+	public void onEvent(Event event) throws Exception {
+		final Map<String, String> data = (Map<String, String>) event.getData();
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/script/DefaultScriptErrorListener.java:33]: ScriptErrorEvent

File: zk/src/archive/web/js/zk/au.ts [hunk 4]
```diff
@@ -1312,7 +1312,16 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
+			if (scriptErrorHandler)
```
Comment (DevChu) [inline zk/src/archive/web/js/zk/au.ts:1322]: why if aync

File: zk/src/archive/web/js/zk/au.ts [hunk 5]
```diff
@@ -1312,6 +1312,14 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
 	 * @param String script the JavaScript code snippet to execute
 	 */
 	script: function (script) {
+		let scriptErrorHandler = zk.scriptErrorHandler;
```
Comment (DevChu) [inline zk/src/archive/web/js/zk/au.ts:1315]: check stack trace (try async) -> from Clients.evalJavascript

Comment (DevChu) [reply zk/src/archive/web/js/zk/au.ts:1315]: get all error -> separate global or ZK

File: zk/src/archive/web/js/zk/mount.ts [hunk 6]
```diff
@@ -837,6 +837,24 @@ jq(function () {
 	.on((document.hidden !== undefined ? '' : zk.vendor_) + 'visibilitychange', function (evt) {
 		zAu._onVisibilityChange();
 	});
+
+	if (zk.scriptErrorHandlerEnabled && !zk.ie9_) {
+		zk.scriptErrorHandler = function (evt) {
+			let errorMsg = evt.originalEvent.message,
+				stack = evt.originalEvent.error.stack,
+				checkFunctionList = (zk.ff || zk.safari) ? ['DOMEval@', 'globalEval@', 'script@', 'doProcess@', 'doCmdsNow@']
```
Comment (DevChu) [inline zk/src/archive/web/js/zk/mount.ts:845]: remove "@" and " ("

File: zk/src/org/zkoss/zk/ui/event/ScriptErrorEvent.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @author jameschu
+ * @since 9.6.0
+ */
+public class ScriptErrorEvent extends Event {
+	private final String _message;
+	private final String _stack;
+
+	public ScriptErrorEvent(String name, String message, String stack) {
+		super(name, new AbstractComponent()); //put dummy target
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/event/ScriptErrorEvent.java:31]: change to static?

File: zktest/src/archive/test2/F96-ZK-4914.zul [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+			<library-property>
+				<name>org.zkoss.ui.script.ScriptErrorListener.class</name>
+				<value>F96_ZK_4914_ScriptErrorListener</value>
+			</library-property>
+		2. Click 'Click and error' button and see "Clients.evalJavaScript Error!" in zk log
+		3. Click 'Click and no error' button twice and click 'Click and error', and you should only see "Clients.evalJavaScript Error!" once in zk log
+		4. Click 'Click and client error' button, and you should not see "Clients.evalJavaScript Error!" in zk log
+	]]></label>
+	<div>
+		<button id="btn" onClick='Clients.evalJavaScript("zk.log(\"log\");")'>Click and see zk log </button>
```
Comment (DevChu) [inline zktest/src/archive/test2/F96-ZK-4914.zul:27]: add two setTimeout case

File: zk/src/archive/web/js/zk/au.ts [hunk 9]
```diff
@@ -1312,6 +1312,11 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
+		if (scriptErrorHandler && !zk.scriptErrorHandlerRegistered) {
+			zk.scriptErrorHandlerRegistered = true;
```
Comment (DevChu) [inline zk/src/archive/web/js/zk/au.ts:1317]: off -> one (don't use flag)

File: @types/zk/zk.d.ts [hunk 10]
```diff
@@ -688,6 +688,9 @@ declare namespace zk {
         resendTimeout: number;
         resourceURI: string;
         safari?: boolean;
+        scriptErrorHandler?: (evt) => void;
+        scriptErrorHandlerEnabled?: boolean;
+        scriptErrorHandlerRegistered?: boolean;
```
Comment (DevChu) [inline @types/zk/zk.d.ts:693]: remove

**Review by scribetw — CHANGES_REQUESTED** [review]
Is dummy component static?

**Review by scribetw — APPROVED** [review]

**Pull request merged**